### PR TITLE
FIX-#340: Fix hangs at low UNIDIST_MPI_SHARED_OBJECT_STORE_THRESHOLD

### LINKED ISSUE #340: MPI backend hangs when setting `UNIDIST_MPI_SHARED_OBJECT_STORE_THRESHOLD` to a lower threshold
Tested with `UNIDIST_MPI_SHARED_OBJECT_STORE_THRESHOLD=100 UNIDIST_CPUS=8 UNIDIST_MPI_SHARED_OBJECT_STORE=True mpiexec -n 1 python -m pytest unidist/test/test_task.py`.

### PR DESCRIPTION
**Pull request by Retribution98**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://unidist.readthedocs.io/en/latest/developer/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 .`
- [x] passes `black .`
- [x] signed commit with `git commit -s` 
- [x] Resolves #340 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: unidist/core/backends/mpi/core/shared_object_store.py [hunk 1]
```diff
@@ -154,30 +151,90 @@ def _allocate_shared_memory(self):
...
             info=info,
         )
         self.shared_buffer, _ = self.win.Shared_query(communication.MPIRank.MONITOR)
 
-        # Service shared memory is allocated only once by the monitor process
-        self.service_info_max_count = (
-            self.shared_memory_size // MpiSharedObjectStoreThreshold.get()
-        ) * self.INFO_SIZE
+        self.service_info_max_count = self.service_memory_size // (
+            self.INFO_SIZE * MPI.BYTE.size
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/shared_object_store.py:236]:
```suggestion
            self.INFO_SIZE * MPI.LONG.size
```

File: unidist/core/backends/mpi/core/shared_object_store.py [hunk 2]
```diff
@@ -154,30 +151,90 @@ def _allocate_shared_memory(self):
...
+            )
+
+        if self.service_memory_size > allowed_memory_size:
+            raise ValueError(
+                "Memory for shared service storage cannot be allocated because the value set to `MpiSharedServiceMemory` exceeds the available memory."
+            )
+
+        if self.shared_memory_size + self.service_memory_size > allowed_memory_size:
+            raise ValueError(
+                "The sum of the `MpiSharedObjectStoreMemory` and `MpiSharedServiceMemory` values is greater than the amount of memory that exists."
```
Comment (YarShev) [inline unidist/core/backends/mpi/core/shared_object_store.py:219]:
```suggestion
                "The sum of the `MpiSharedObjectStoreMemory` and `MpiSharedServiceMemory` values is greater than the available amount of memory."
```
Put this message in multiple lines but not in one.

File: unidist/core/backends/mpi/core/shared_object_store.py [hunk 3]
```diff
@@ -154,30 +151,90 @@ def _allocate_shared_memory(self):
...
+                if B > 99 * C:
+                    self.shared_memory_size = (A * B) // (B + C)
+                    self.service_memory_size = (A * C) // (B + C)
+                else:
+                    self.shared_memory_size = int(0.99 * A)
+                    self.service_memory_size = int(0.01 * A)
+
+        if self.shared_memory_size > allowed_memory_size:
+            raise ValueError(
+                "Memory for shared object storage cannot be allocated because the value set to `MpiSharedObjectStoreMemory` exceeds the available memory."
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/shared_object_store.py:209]: ok

File: unidist/core/backends/mpi/core/shared_object_store.py [hunk 4]
```diff
@@ -154,30 +151,90 @@ def _allocate_shared_memory(self):
...
+            )
+
+        if self.service_memory_size > allowed_memory_size:
+            raise ValueError(
+                "Memory for shared service storage cannot be allocated because the value set to `MpiSharedServiceMemory` exceeds the available memory."
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/shared_object_store.py:214]: Ok

File: unidist/core/backends/mpi/core/shared_object_store.py (hunk 1 above)
Comment (Retribution98) [reply unidist/core/backends/mpi/core/shared_object_store.py:236]: It has been fixed in another way.

File: unidist/core/backends/mpi/core/shared_object_store.py [hunk 5]
```diff
@@ -154,30 +151,90 @@ def _allocate_shared_memory(self):
         """
         mpi_state = communication.MPIState.get_instance()
 
-        # Use only 95% of available shared memory because
-        # the rest is needed for intermediate shared buffers
-        # handled by MPI itself for communication of small messages.
-        # Shared memory is allocated only once by the monitor process.
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply unidist/core/backends/mpi/core/shared_object_store.py:160]: It has been back.

**Review by YarShev — APPROVED** [review]

**Pull request merged**